oak_idl_gen_services/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+    let generated_schema = generate_from_bytes(&schema_bytes).unwrap();
+    let out_file = format!(
+        "{}/{}_services.rs",
+        std::env::var("OUT_DIR").unwrap(),
+        file_prefix
+    );
+    std::fs::write(&out_file, generated_schema).unwrap();
+}
+
+/// Generate the Rust file from the input flatbuffer file.
+///
+/// In general, each flatbuffer file may have zero or more `rpc_service` instances defined.
+fn generate_from_bytes(schema_bytes: &[u8]) -> anyhow::Result<String> {
+    let schema = reflection_generated::reflection::root_as_schema(schema_bytes)
+        .context("could not parse schema")?;
+    Ok(schema
+        .services()
+        .context("could not find any service definition")?
+        .iter()
+        .map(|s| generate_service(&s))
+        .collect::<Result<Vec<_>, _>>()?
+        .into_iter()
+        .intersperse("\n".to_string())
+        .collect())
+}
+
+/// Generate the Rust objects from the input [`Service`] instance, corresponding to an `rpc_service`
+/// entry.
+fn generate_service(service: &Service) -> anyhow::Result<String> {
+    let mut lines = Vec::new();
+    lines.extend(vec![
+        format!("// Generated file, do not edit."),
+        format!(""),
+        format!(
+            "pub struct {}<T: oak_idl::Transport> {{",
+            client_name(service)
+        ),
+        format!("    transport: T"),
+        format!("}}"),
+        format!(""),
+        format!("impl <T: oak_idl::Transport>{}<T> {{", client_name(service)),
+        format!("    pub fn new(transport: T) -> Self {{"),
+        format!("        Self {{"),
+        format!("            transport"),
+        format!("        }}"),
+        format!("    }}"),
+    ]);
+    lines.extend(
+        service
+            .calls()
+            .context("could not find any call definitions")?
+            .iter()
+            .map(|c| generate_client_method(&c))
+            .collect::<Result<Vec<_>, _>>()
+            .context("could not generate client method")?
+            .into_iter()
+            .flatten(),
+    );
+    lines.extend(vec![format!("}}"), format!("")]);
+    lines.extend(vec![
+        format!("pub struct {}<S> {{", server_name(service)),
+        format!("    service: S"),
+        format!("}}"),
+        format!(""),
+        format!(
+            "impl <S: {}> oak_idl::Transport for {}<S> {{",
+            service_name(service),
+            server_name(service)
+        ),
+        format!("    fn invoke(&mut self, request_message: oak_idl::TransportMessage) -> Result<oak_idl::TransportMessage, oak_idl::TransportError> {{"),
+        format!("        let response_header = oak_idl::Header {{"),
+        format!("            transaction_id: request_message.header.transaction_id,"),
+        format!("            method_id: request_message.header.method_id,"),
+        format!("        }};"),
+        format!("        match request_message.header.method_id {{"),
```

Comment:
After reading the flatbuffer docs I'm a little surprised that there isn't an invoke based transport abstraction like this incorporated into flatbuffer itself. 

Instead their approach seems to be to generated gRPC code directly (only available in a select few languages). With an invoke based abstraction layer they could implement transport agnostic stubs once for each language and then allowing users to swap in a transport protocol that works for them.